provisioner/data.js

```diff
@@ -265,83 +387,51 @@ WorkerType.createLaunchSpec = function (region, instanceType, worker, keyPrefix,
   var regionSpecificKeys = [
     'ImageId', // AMI IDs (ImageId) are created and are different per-region
   ];
-
-  // Find the region overwrites object
-  var regionOverwrites;
-  worker.regions.forEach(function (r) {
-    if (r.region === region) {
-      assert(!regionOverwrites, 'regions must be unique');
-      regionOverwrites = r.overwrites;
-    }
-  });
-  assert(regionOverwrites);
-
-  // Find the instanceType overwrites object
-  var typeOverwrites;
-  worker.instanceTypes.forEach(function (t) {
-    if (t.instanceType === instanceType) {
-      assert(!typeOverwrites, 'instanceTypes must be unique');
-      typeOverwrites = t.overwrites;
-    }
-  });
-  assert(typeOverwrites);
-
   // Check for type specific keys in the general keys and region keys
   typeSpecificKeys.forEach(function (key) {
-    if (worker.launchSpecification[key]) {
+    if (worker.launchSpec[key]) {
       throw new Error(key + ' is type specific, not general');
     }
-    if (regionOverwrites[key]) {
+    if (regionOverwriteObjects.launchSpec[key]) {
       throw new Error(key + ' is type specific, not type specific');
     }
   });
 
   // Check for region specific keys in the general and type keys
   regionSpecificKeys.forEach(function (key) {
-    if (worker.launchSpecification[key]) {
+    if (worker.launchSpec[key]) {
       throw new Error(key + ' is region specific, not general');
     }
-    if (typeOverwrites[key]) {
+    if (instanceTypeOverwriteObjects.launchSpec[key]) {
       throw new Error(key + ' is type specific, not region specific');
     }
   });
 
-  // InstanceTypes should be unique inside of a worker
-  var allInstanceTypes = [];
-  worker.instanceTypes.forEach(function (i) {
-    if (allInstanceTypes.includes(i.instanceType)) {
-      throw new Error(worker.workerType + ' has duplicated instanceType ' + i.instanceType);
-    } else {
-      allInstanceTypes.push(i);
-    }
+  var config = {};
+
+  // Do the cascading overwrites of the object things
+  ['launchSpec', 'userData', 'secrets'].forEach(x => {
+    config[x] = lodash.cloneDeep(worker[x] || {});
+    lodash.assign(config[x], regionOverwriteObjects[x]);
+    lodash.assign(config[x], instanceTypeOverwriteObjects[x]);
```

Comment:
Neat, note that you can give more than one to `_.assign`, hence:
```js
config[x] = _.cloneDeep(_.assign({}, worker[x], regionOverwriteObjects[x], instanceTypeOverwriteObjects[x]));
```
Note, cloneDeep is only need if you actually modify the properties inside `config[x]` later...